### PR TITLE
Introduce ReplicaEngine model

### PR DESCRIPTION
**Pull request by DaveCTurner**
This models how indexing and deletion operations are handled on the replica,
including the optimisations for append-only operations and the interaction with
Lucene commits and the version map.

NB this version is based on [`master`](https://github.com/elastic/elasticsearch/commit/70f67b17dd1995077af314c4f3fa67d5b337077f) and exhibits a bug. I'm opening this for comments on approach etc and will followup with changes as per PRs.

### CODE REVIEW

File: ReplicaEngine/tla/ReplicaEngine.tla [hunk 1]
```diff
@@ -0,0 +1,817 @@
...
+       RETRY_ADD: A retry of a write that does involve an internally-generated
+       document ID.
+       Both can occur at any seqno. *)
+    \cup  { [type |-> UPDATE, seqno |-> seqno, content |-> content]
+          : seqno   \in 1..request_count
+          , content \in DocContent
+          }
+    (* DELETE *)
+    \cup  { [type |-> DELETE, seqno |-> seqno]
+          : seqno \in 1..request_count
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply ReplicaEngine/tla/ReplicaEngine.tla]: Probably not. But I think I've found a modelling bug involving 3 DELETEs and then an UPDATE (2 DELETEs was not enough) so requiring a preceding UPDATE would make a much bigger model.

File: ReplicaEngine/tla/ReplicaEngine.tla [hunk 2]
```diff
@@ -0,0 +1,817 @@
...
+  while replication_requests /= {} do
+    with replication_request \in replication_requests do
+        req := replication_request;
+        replication_requests := replication_requests \ {replication_request};
+    end with;
+    
+    if req.type = DELETE
+    then
+        (* planDeletionAsNonPrimary *)
+
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply ReplicaEngine/tla/ReplicaEngine.tla]: I left that out because it's not in the code.

[the remark this answers is not in the document]
Comment (bleskes) [reply ReplicaEngine/tla/ReplicaEngine.tla]: see https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java#L1040

[the remark this answers is not in the document]
Comment (DaveCTurner) [reply ReplicaEngine/tla/ReplicaEngine.tla]: Ohhhh. Outside the lock, so I missed it, d'oh.

[the remark this answers is not in the document]
Comment (DaveCTurner) [reply ReplicaEngine/tla/ReplicaEngine.tla]: Fixed in 0ceddd1

File: ReplicaEngine/tla/ReplicaEngine.tla (hunk 1 above)
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply ReplicaEngine/tla/ReplicaEngine.tla]: Ok, @ywelsch gives a scenario where an UPDATE and a DELETE are replicated by the primary which then crashes, and the new primary only received the DELETE so replaces the UPDATE with a no-op.

File: ReplicaEngine/tla/ReplicaEngine.tla [hunk 3]
```diff
@@ -0,0 +1,832 @@
...
+                plan                     := "processButSkipLucene";
+                currentNotFoundOrDeleted := FALSE;
+                useLuceneUpdateDocument  := FALSE;
+                indexIntoLucene          := FALSE;
+            else
+                if versionMap_isUnsafe
+                then
+                    (* Perform a Lucene refresh *)
+                    AwaitRefreshOnIndex: \* Label here to allow for other concurrent activity
+                    await lucene.buffer = <<>>;
```
Comment (ywelsch) [inline ReplicaEngine/tla/ReplicaEngine.tla]: same comment as above for delete, do we need to do enforceSafeAccess here?

File: ReplicaEngine/tla/ReplicaEngine.tla [hunk 4]
```diff
@@ -0,0 +1,832 @@
...
+          : content \in DocContent
+          }
+    (* RETRY_ADD: A retry of a write that does involve an internally-generated
+       document ID. *)
+    \cup  { [type |-> RETRY_ADD, seqno |-> seqno, content |-> content, autoIdTimeStamp |-> DocAutoIdTimestamp]
+          : seqno   \in 1..request_count
+          , content \in DocContent
+          }
+    (* UPDATE: A write that does not involve an internally-generated document ID. *)
+    \cup  { [type |-> UPDATE, seqno |-> seqno, content |-> content]
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply ReplicaEngine/tla/ReplicaEngine.tla]: Nice.

File: ReplicaEngine/tla/ReplicaEngine.tla [hunk 5]
```diff
@@ -0,0 +1,832 @@
...
+        end with;
+    end while;
+end process;
+
+(* Lucene refreshes can happen at any time *)
+process LuceneProcess = "ReplicaLucene"
+variables
+    lucene =
+        [ document |-> NULL
+        , buffer   |-> <<>>
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply ReplicaEngine/tla/ReplicaEngine.tla]: Ok, done.

File: ReplicaEngine/tla/ReplicaEngine.tla [hunk 6]
```diff
@@ -0,0 +1,832 @@
...
+            (* OP_STALE_OR_EQUAL *)
+            plan             := "processButSkipLucene";
+            deleteFromLucene := FALSE;
+            currentlyDeleted := FALSE;
+        else
+            if versionMap_isUnsafe
+            then
+                (* Perform a Lucene refresh *)
+                AwaitRefreshOnDelete: \* Label here to allow for other concurrent activity
+                await lucene.buffer = <<>>;
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply ReplicaEngine/tla/ReplicaEngine.tla]: Good catch. Added.

File: ReplicaEngine/tla/ReplicaEngine.tla (hunk 3 above)
Comment (DaveCTurner) [reply ReplicaEngine/tla/ReplicaEngine.tla]: Yep, good catch, added.

File: ReplicaEngine/tla/ReplicaEngine.tla [hunk 7]
```diff
@@ -0,0 +1,832 @@
...
+        await lucene.buffer /= <<>>;
+        lucene := [lucene EXCEPT
+            !.document = ApplyBufferedOperations(lucene.buffer, @),
+            !.buffer   = <<>>
+            ];
+            
+        LuceneUpdateVersionMap: (* TODO needs an invariant saying that the VM is >= Lucene and also contains the buffered ops *)
+                        
+        versionMap_isUnsafe := FALSE;
+        versionMap_needsSafeAccess := FALSE;
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply ReplicaEngine/tla/ReplicaEngine.tla]: It makes it way less likely, but I don't think it eliminates it. You just need more concurrent activity and more refreshes in order to clear it. I think it's worth modelling more correctly anyway.

[the remark this answers is not in the document]
Comment (DaveCTurner) [reply ReplicaEngine/tla/ReplicaEngine.tla]: > I wonder if this addresses the oddity we have seen after you added the fixes to the model.

The oddity that the model found is in fact another desync bug that's reproducible in Elasticsearch itself.

**Comment by DaveCTurner** [comment]
@bleskes said:

> In terms of invariants in the model, I think we should add the following:
> 
> 1) Assert that at any given moment the version map is a good representation of the Lucene, when combine with the indexing buffer. This is not so straight forward - we need to make sure that if you ask the version map when it thinks it's safe, it's answer is always consistent with what Lucene would say if you hit refresh.
>
> 2) The enforceSafeAccess flag is an optimization to make sure we don't hit refresh storms. The correctness should never depend on it. I wonder if we should adapt the model to flip it on will (rather than connect it to two refresh cycles).

Until this model is in a state where it passes the model checker I'd prefer not to add artificial invariants or transitions. Both of those ideas would have reduced the number of steps it took to find a failure, but would have made it much harder to answer the question of whether it was a genuine bug (i.e. actually could lead to a desync) or not.

**Comment by DaveCTurner** [comment]
I pushed fixes that model elastic/elasticsearch#28787 and elastic/elasticsearch#28790 as well as @bleskes proposal for the third issue.

I also dropped the label from the middle of the refresh loop as it was yielding spurious failures because of not modelling both the current and old maps.

This model now passes the model checker - takes about a minute to go through ~6M states on my laptop.

**Comment by DaveCTurner** [comment]
Incidentally, trying to reproduce the third desync bug in the test harness fails because when assertions are enabled we always call `removeTombstonesUnderLock`.

File: ReplicaEngine/tla/ReplicaEngine.tla [hunk 8]
```diff
@@ -0,0 +1,834 @@
...
+        localCheckPoint := localCheckPoint + 1;
+    end while;
+end process
+
+(* The process that consumes replication requests for a particular document ID, which
+   are processed in series because of the lock in the version map. *)
+process ConsumerProcess = "Consumer"
+variables
+    maxUnsafeAutoIdTimestamp \in {0, DocAutoIdTimestamp - 1, DocAutoIdTimestamp, DocAutoIdTimestamp + 1},
+    maxSeqNoOfNonAppendOnlyOperations \in {0, 2, request_count + 1},
```
Comment (ywelsch) [inline ReplicaEngine/tla/ReplicaEngine.tla]: why only `0, 2, request_count + 1`? How about starting with 0, and then (similarly as you have done for modeling other concurrent requests for different documents) model a separate process that increases this.

File: ReplicaEngine/tla/ReplicaEngine.tla [hunk 9]
```diff
@@ -0,0 +1,834 @@
...
+        if /\ req.type = ADD
+           /\ maxUnsafeAutoIdTimestamp < req.autoIdTimeStamp
+           /\ maxSeqNoOfNonAppendOnlyOperations < req.seqno \* PR #28787
+        then
+            plan                     := "optimisedAppendOnly";
+            currentNotFoundOrDeleted := TRUE;
+            useLuceneUpdateDocument  := FALSE;
+            indexIntoLucene          := TRUE;
+        else
+            if FALSE = (req.type \in {ADD, RETRY_ADD})
```
Comment (ywelsch) [inline ReplicaEngine/tla/ReplicaEngine.tla]: I think that `req.type \notin {ADD, RETRY_ADD}` would read nicer.

File: ReplicaEngine/tla/ReplicaEngine.tla (hunk 8 above)
Comment (DaveCTurner) [reply ReplicaEngine/tla/ReplicaEngine.tla]: I was battling state space explosion. However, this doesn't seem to be the source of it, so I pushed 36454d7dc56c6ff9c90332f7b2ede0c729cf43f7.

File: ReplicaEngine/tla/ReplicaEngine.tla (hunk 9 above)
Comment (DaveCTurner) [reply ReplicaEngine/tla/ReplicaEngine.tla]: Me too. I pushed a643c14f8aebdc60730d79f4bfaf8de932fb5830

**Comment by DaveCTurner** [comment]
I missed something: we're not currently dealing with duplicated messages. I pushed 8cd6ce93da8ba5687fd0ffafb872354fd1ae6633. It's now much slower.

**Pull request merged**